Fix reset logic in nav2_behavior_tree control nodes

naiveHobo opened the pull request:
Signed-off-by: Sarthak Mittal <sarthakmittal2608@gmail.com>

<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #2002, #2003  |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | TurtleBot3 Gazebo Simulation |

---

## Description of contribution in a few bullet points

* Fixes `RecoveryNode` logic which had a few errors when halting children
* Fixes halting logic in `PipelineSequenceNode`
* Updates unit tests to reflect the changes

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/plugins/control/round_robin_node.hpp:
```diff
@@ -30,7 +30,8 @@ namespace nav2_behavior_tree
  *
  * If the current child return failure, the next child is ticked and if the last child returns
- * failure, the first child is ticked and the cycle continues until a child returns success
+ * failure, the first child is ticked and the cycle continues until a child returns success.
```
Please add this change to the galactic migration guide so that people know we changed behavior

SteveMacenski commented on nav2_behavior_tree/plugins/control/round_robin_node.cpp:
```diff
@@ -74,7 +73,11 @@ BT::NodeStatus RoundRobinNode::tick()
   }
 
-  halt();
+  // reset
```
The `halt()` function in RoundRobbinNode is no longer called. Should it be removed, or should that be what's updated? Seems like they're pretty much the same except `   ControlNode::haltChildren(); ` is `ControlNode::halt();`. I think you should just change them in that function

naiveHobo replied:
The design pattern I noticed in the behavior tree library was the for control nodes `halt()` is never called within the `tick()` function. `halt()` calls `haltChildren()` internally but it also sets the status of the current node back to `IDLE`. Within the `tick()` function, the control node is expected to halt its children on failure/success but not set itself back to `IDLE`. Calling the control node's `halt()` function is its parent's job and the `halt()` function is only expected to reset the node's internal variables.

SteveMacenski replied:
We override `halt` in this class and you removed the only use of it in the code, I think that's why the code coverage drops so much. The only difference you added and the function as written is 1 line that we could just update?

crdelsey commented on nav2_behavior_tree/include/nav2_behavior_tree/plugins/control/round_robin_node.hpp:
```diff
@@ -42,8 +43,9 @@ namespace nav2_behavior_tree
  * | FAILURE | SUCCESS |  IDLE   |  - B returns SUCCESS, so RoundRobin halts all children and
- *                                  - returns SUCCESS, next iteration will tick C.
- * | RUNNING |  IDLE   | FAILURE |  - C returns FAILURE, so RoundRobin circles and ticks A.
+ *                                  - returns SUCCESS, next iteration will tick B again.
```
I'm not sure that this change in behavior is really what we want. For instance, in `navigate_w_replanning_and_round_robin_recovery.xml` in this block
```
<RoundRobin name="GlobalPlannerRecoveryActions">
     <ClearEntireCostmap name="ClearGlobalCostmap-Context" service_name="global_costmap/clear_entirely_global_costmap"/>
    <Wait wait_duration="2"/>
</RoundRobin>
```
If `clearcostmap` fails, we'd advance to `wait` on the next tick. `wait` seems unlikely to ever fail, so we'd be stuck on that recovery action forever

naiveHobo replied:
You're right, going to revert this

Jconn commented:
I verified on my gazebo setup that https://github.com/ros-planning/navigation2/issues/2003 is fixed with this pr.

naiveHobo commented:
@SteveMacenski the coverage changes don't seem to be related to the changes in this PR now

SteveMacenski commented:
Yeah, fair enough

naiveHobo commented:
> I see you reverted a bunch of stuff, can you tell us a little bit about what you just changed and why?

I reverted the `RoundRobin` logic change I introduced earlier after @crdelsey's [comment](https://github.com/ros-planning/navigation2/pull/2004#discussion_r497882941). This is unrelated to the linked tickets so @Jconn's `PipelineSequence` issue should still be resolved. The `RecoveryNode` is changed to call `halt()` instead of manually resetting the node on success/failure based on @SteveMacenski's [comment](https://github.com/ros-planning/navigation2/pull/2004#discussion_r497755304).

This PR should be okay to merge now. Both the linked tickets are being resolved.

SteveMacenski commented:
Ok, I want @crdelsey and @mkhansenbot to take another look and approve things, the control flow nodes are probably the thing I'm least comfortable with in the code base and having a sanity check from the original authors is good.

crdelsey reviewed (approved):
This looks good to me now, though I think more was reverted than needed to be.

mjeronimo commented:
@mkhansenbot  Sure, I had a look and the changes look good to me.
